src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala

```diff
@@ -0,0 +1,619 @@
+    val key4: Array[Byte] = Hex.decode("123500")
+    val trie = EmptyTrie.put(key1, key1).put(key2, key2).put(key3, key3).put(key4, key4)
+    val wrongSource = HashMapDataSource().update(
+      rootHash = Array.emptyByteArray,
+      toRemove = Seq(),
+      toUpdate = Seq(trie.getRootHash -> trie.dataSource.get(trie.getRootHash).get)
+    )
+    val trieWithWrongSource = MerklePatriciaTrie[Array[Byte], Array[Byte]](trie.getRootHash, wrongSource, hashFn)
+    val trieAfterDelete = Try {
+      trieWithWrongSource.remove(key1)
+    }
+    assert(trieAfterDelete.isFailure)
+  }
+
+  test("Get in an empty trie"){
+    val obtained = EmptyTrie.get(Hex.decode("1234"))
+    assert(obtained.isEmpty)
+  }
+
+  test("Insert with an empty key in a branch node"){
+    val key1: Array[Byte] = Hex.decode("00")
+    val key2: Array[Byte] = Hex.decode("f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val key3: Array[Byte] = Hex.decode("")
+    val val3: Array[Byte] = Hex.decode("1234")
+    val trie = trieWithBranch.put(key3, val3)
+    val key1Get = trie.get(key1)
+    val key2Get = trie.get(key2)
+    val key3Get = trie.get(key3)
+    assert(key1Get.isDefined && (key1Get.get sameElements key1))
+    assert(key2Get.isDefined && (key2Get.get sameElements key2))
+    assert(key3Get.isDefined && (key3Get.get sameElements val3))
+  }
+
+  test("Remove of a key (not in trie) whose value should be in a branch node"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  test("Remove of a key (not in trie) that should be in the child of a BranchNode that is not present"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11a0"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  /* IODB tests */
+  test("Simple test with IODB as Source") {
+    //create temporary dir
+    val dir = File.createTempFile("iodb", "iodb")
+    dir.delete()
+    dir.mkdir()
+
+    //open new store
+    val dataSource = new IodbDataSource(new LSMStore(dir = dir, keySize = 32))
+    val emptyTrie = MerklePatriciaTrie[Int, Int](dataSource, hashFn)
+    val trieWithOneElement = emptyTrie.put(1, 5)
+    val obtained = trieWithOneElement.get(1)
+    assert(obtained.isDefined)
+    assert(obtained.get == 5)
+    val trieAfterDelete = trieWithOneElement.remove(1)
+    val obtainedAfterDelete = trieAfterDelete.get(1)
+    assert(obtainedAfterDelete.isEmpty)
+  }
+
+  ignore("IODB test - Insert of the first 5000 numbers hashed and then remove half of them"){
```

Comment:
@rtkaczyk no, `sbt test`won't run all tests because it's filtering long running ones by default `Tests.Argument("-l", "LongRunningTest")`. I really don't have a preferred way to do this (tags vs folders) so let's decide it together. I'm open to the one that we all like most and have no problem to update the code accordingly.